File: python/packages/jobs-client/jobs_client/observability.py
"""Metrics + heartbeat-backed health server for the warmworker.

The warmworker (`jobs_client.warm_consumer`) is a pure client and cannot import
the service internals, so this DUPLICATES the cold dispatcher's observability
module on the consumer's own prometheus registry. The metric NAMES match
jobs-service's so Prometheus aggregates across both targets. Only the
`connection_up` gauge carries an explicit `endpoint` label — it's a per-process
liveness signal that must distinguish warm vs cold within a single query; the
counters/histograms rely on Prometheus's automatic per-target `job`/`instance`
labels instead, so they deliberately don't add a process label. /healthz is
heartbeat-checked because the HTTP server runs in its own thread and answers
/metrics even if the asyncio loop hangs."""

from __future__ import annotations

import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from prometheus_client import CollectorRegistry, Counter, Gauge, Histogram, generate_latest

REGISTRY = CollectorRegistry()

queue_consume_total = Counter(
    "jobs_queue_consume_total", "Queue consumer outcomes",
    labelnames=("job_name", "dispatch_mode", "outcome"),
    registry=REGISTRY,
)
queue_attempt_count = Histogram(
    "jobs_queue_attempt_count", "Attempt number at terminal disposition",
    labelnames=("job_name", "dispatch_mode"),
    registry=REGISTRY,
)
queue_connection_up = Gauge(
    "jobs_queue_connection_up", "1 if the queue connection is up, 0 if reconnecting",
    labelnames=("endpoint",),
    registry=REGISTRY,
)
queue_pending_messages = Gauge(
    "jobs_queue_pending_messages", "Messages waiting in each work queue",
    labelnames=("job_name", "dispatch_mode"),
    registry=REGISTRY,
)
queue_dlq_messages = Gauge(
    "jobs_queue_dlq_messages", "Messages in each dead-letter queue",
    labelnames=("job_name", "dispatch_mode"),
    registry=REGISTRY,
)
warm_handler_duration_seconds = Histogram(
    "jobs_warm_handler_duration_seconds", "Time inside the warm handler callable",
    labelnames=("job_name", "outcome"),
    registry=REGISTRY,
)


def render() -> bytes:
    return generate_latest(REGISTRY)


class Heartbeat:
    """Monotonic last-progress timestamp. Bumped by the async loop, read by the
    HTTP thread; a Lock keeps the cross-thread read/write tidy."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._last = time.monotonic()

    def beat(self) -> None:
        with self._lock:
            self._last = time.monotonic()

    def seconds_since(self) -> float:
        with self._lock:
            return time.monotonic() - self._last


def start_metrics_server(
    *, port: int, heartbeat: Heartbeat, staleness_seconds: int,
) -> ThreadingHTTPServer:
    """Daemon-thread HTTP server serving /metrics and /healthz on `port`.
    /healthz is 200 while the heartbeat is fresh, 503 once stale."""

    class _Handler(BaseHTTPRequestHandler):
        def do_GET(self):  # noqa: N802 — stdlib-mandated name
            if self.path == "/metrics":
                body = render()
                self.send_response(200)
                self.send_header("Content-Type", "text/plain; version=0.0.4")
                self.end_headers()
                self.wfile.write(body)
            elif self.path == "/healthz":
                stale = heartbeat.seconds_since() >= staleness_seconds
                self.send_response(503 if stale else 200)
                self.send_header("Content-Type", "text/plain")
                self.end_headers()
                self.wfile.write(b"stale\n" if stale else b"ok\n")
            else:
                self.send_response(404)
                self.end_headers()

        def log_message(self, *args):  # silence per-request stderr noise
            pass

    server = ThreadingHTTPServer(("0.0.0.0", port), _Handler)
    server.daemon_threads = True
    threading.Thread(target=server.serve_forever, name="metrics-http", daemon=True).start()
    return server

File: dotnet/src/BrutalSystems.Jobs.Worker/Observability.cs
using System.Net;
using System.Net.Sockets;
using System.Runtime.CompilerServices;
using System.Text;
using Prometheus;

[assembly: InternalsVisibleTo("BrutalSystems.Jobs.Worker.Tests")]

namespace BrutalSystems.Jobs.Worker;

/// <summary>Monotonic last-progress timestamp. Bumped by the consumer loop, read by the
/// health endpoint. Port of jobs_client.observability.Heartbeat.</summary>
public sealed class Heartbeat
{
    private readonly TimeProvider _clock;
    private readonly Lock _gate = new();
    private DateTimeOffset _last;

    public Heartbeat(TimeProvider? clock = null)
    {
        _clock = clock ?? TimeProvider.System;
        _last = _clock.GetUtcNow();
    }

    public void Beat() { lock (_gate) _last = _clock.GetUtcNow(); }
    public TimeSpan SinceLast() { lock (_gate) return _clock.GetUtcNow() - _last; }
}

/// <summary>Warm-consumer metrics on a private registry. Metric NAMES match jobs-service's so
/// Prometheus aggregates across targets. Port of jobs_client.observability.</summary>
public sealed class JobsMetrics
{
    private readonly CollectorRegistry _registry = Metrics.NewCustomRegistry();

    public Counter QueueConsumeTotal { get; }
    public Histogram QueueAttemptCount { get; }
    public Gauge QueueConnectionUp { get; }
    public Gauge QueuePendingMessages { get; }
    public Gauge QueueDlqMessages { get; }
    public Histogram WarmHandlerDurationSeconds { get; }

    public JobsMetrics()
    {
        var f = Metrics.WithCustomRegistry(_registry);
        QueueConsumeTotal = f.CreateCounter("jobs_queue_consume_total", "Queue consumer outcomes",
            new CounterConfiguration { LabelNames = ["job_name", "dispatch_mode", "outcome"] });
        QueueAttemptCount = f.CreateHistogram("jobs_queue_attempt_count", "Attempt number at terminal disposition",
            new HistogramConfiguration { LabelNames = ["job_name", "dispatch_mode"] });
        QueueConnectionUp = f.CreateGauge("jobs_queue_connection_up", "1 if the queue connection is up, 0 if reconnecting",
            new GaugeConfiguration { LabelNames = ["endpoint"] });
        QueuePendingMessages = f.CreateGauge("jobs_queue_pending_messages", "Messages waiting in each work queue",
            new GaugeConfiguration { LabelNames = ["job_name", "dispatch_mode"] });
        QueueDlqMessages = f.CreateGauge("jobs_queue_dlq_messages", "Messages in each dead-letter queue",
            new GaugeConfiguration { LabelNames = ["job_name", "dispatch_mode"] });
        WarmHandlerDurationSeconds = f.CreateHistogram("jobs_warm_handler_duration_seconds", "Time inside the warm handler callable",
            new HistogramConfiguration { LabelNames = ["job_name", "outcome"] });
    }

    public string Render()
    {
        // Sync-over-async is fine here: /metrics is scraped rarely and off the hot path.
        using var ms = new MemoryStream();
        _registry.CollectAndExportAsTextAsync(ms).GetAwaiter().GetResult();
        return Encoding.UTF8.GetString(ms.ToArray());
    }

    public (int Status, string Body) Health(int stalenessSeconds, Heartbeat hb)
        => hb.SinceLast() >= TimeSpan.FromSeconds(stalenessSeconds) ? (503, "stale\n") : (200, "ok\n");
}

/// <summary>HttpListener serving /metrics + /healthz. /healthz is heartbeat-checked so an
/// event-loop stall trips k8s liveness even while /metrics still answers.</summary>
public static class MetricsServer
{
    /// <summary>Allows tests to read the OS-assigned port when port 0 is passed to Start.</summary>
    internal interface IHavePort
    {
        int Port { get; }
    }

    public static IDisposable Start(int port, JobsMetrics metrics, Heartbeat heartbeat, int stalenessSeconds,
        string? host = null)
    {
        // Default to all interfaces — k8s liveness probes the pod IP, not loopback. The `+` strong
        // wildcard needs a urlacl on Windows; a Windows dev can pass METRICS_HOST=localhost instead.
        host ??= Environment.GetEnvironmentVariable("METRICS_HOST") ?? "+";

        // When port 0 is requested, pick an ephemeral port using a TcpListener probe.
        // HttpListener does not support port 0 directly (it uses the HTTP API, not TCP sockets).
        int boundPort = port == 0 ? FindFreePort() : port;

        var listener = new HttpListener();
        listener.Prefixes.Add($"http://{host}:{boundPort}/");
        listener.Start();

        var cts = new CancellationTokenSource();
        _ = Task.Run(async () =>
        {
            while (!cts.IsCancellationRequested)
            {
                HttpListenerContext ctx;
                try { ctx = await listener.GetContextAsync(); }
                catch { break; }
                var h = metrics.Health(stalenessSeconds, heartbeat);
                var (status, body, contentType) = ctx.Request.Url?.AbsolutePath switch
                {
                    "/metrics" => (200, metrics.Render(), "text/plain; version=0.0.4"),
                    "/healthz" => (h.Status, h.Body, "text/plain"),
                    _ => (404, "", "text/plain"),
                };
                var bytes = Encoding.UTF8.GetBytes(body);
                ctx.Response.StatusCode = status;
                ctx.Response.ContentType = contentType;
                await ctx.Response.OutputStream.WriteAsync(bytes);
                ctx.Response.Close();
            }
        });
        return new Stopper(listener, cts, boundPort);
    }

    /// <summary>For test/ephemeral use only — production must always pass a real configured port (never 0), as the probe binds loopback only.</summary>
    private static int FindFreePort()
    {
        // Bind to port 0 to let the OS allocate an ephemeral port, then release it.
        // There is a tiny TOCTOU window, but it is negligible for test use.
        using var tcp = new TcpListener(IPAddress.Loopback, 0);
        tcp.Start();
        int p = ((IPEndPoint)tcp.LocalEndpoint).Port;
        tcp.Stop();
        return p;
    }

    private sealed class Stopper(HttpListener listener, CancellationTokenSource cts, int port)
        : IDisposable, IHavePort
    {
        public int Port => port;
        public void Dispose() { cts.Cancel(); listener.Close(); }
    }
}
